spring-cloud-gcp-starters/spring-cloud-gcp-starter-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/autoconfig/GcpPubsubAutoConfiguration.java

```diff
@@ -112,4 +117,42 @@ public PublisherFactory defaultPublisherFactory(GcpProjectIdProvider projectIdPr
 		return new DefaultPublisherFactory(projectIdProvider, subscriberProvider, channelProvider,
 				credentialsProvider);
 	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	public PubsubAdmin pubsubAdmin(GcpProjectIdProvider projectIdProvider,
+			TopicAdminClient topicAdminClient,
+			SubscriptionAdminClient subscriptionAdminClient) {
+		return new PubsubAdmin(projectIdProvider, topicAdminClient, subscriptionAdminClient);
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	public TopicAdminClient topicAdminClient(CredentialsProvider credentialsProvider) {
+		try {
+			return TopicAdminClient.create(
+					TopicAdminSettings.defaultBuilder()
+							.setCredentialsProvider(credentialsProvider)
+							.build());
+		}
+		catch (IOException ioe) {
+			throw new PubsubException("An error occurred while creating TopicAdminClient.", ioe);
+		}
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	public SubscriptionAdminClient subscriptionAdminClient(
+			CredentialsProvider credentialsProvider) {
+		try {
+			return SubscriptionAdminClient.create(
+					SubscriptionAdminSettings.defaultBuilder()
+							.setCredentialsProvider(credentialsProvider)
```

Comment:
Sorry, didn't mean to not reply here, just went grab lunch in the meantime. :)

`IOException` is thrown because of the channel creation, and maybe other cases, although I'm not sure about the exact circumstances it can happen (internet endpoint is unreachable?). Because of that, I think you do have access to low level resources, like a TCP connection, when you're creating a `Subscription`.

I really like this way because you still have to handle low level resources, but you do that away from the constructor. If you need to handle the `IOException`, you do it away from the constructor. And it should still be easy enough for users to just call `new PubsubAdmin(projectIdProvider, TopicAdminClient.create(), SubscriptionAdminClient.create())`.

Although, to be honest, I'm not opposed to offer that constructor on the side too.